Code/Tools/FBuild/Documentation/docs/functions/settings.html

```diff
@@ -51,6 +51,17 @@ <h1>Settings</h1>
   .AllowDBMigration_Experimental    // (optional) Migrate build state when bff changes to avoid rebuilds (default: false)
                                     // This enables what will eventually be standard behaviour
                                     // WARNING: This feature is experimental.
+  .WorkerListRefreshLimit           // (optional) When not using a static worker list,
+                                    //            max number of seconds to refresh the network share worker list.
+                                    //            default: 300 (5 minutes, in case workers are rebooting)
+                                    //            0  means no retries
+                                    //            -1 means infinite retries
```

Comment:
0 as a special value for "never refresh" is reasonable (to allow this feature to be disabled), but I don't quite understand what "-1 infitinite retries" means in this context

Reply:
I think I understand. I originally thought this was "refresh every 5 minutes".  Is there a reason we should limit searching for new workers? Seems like for a lengthy build, you want to continually search for new workers if some of your workers become unavailable?

Reply:
Yes. WorkerListRefreshLimit is an upper limit on the total time spent looking for new workers.

It is not the worker poll interval. I have the Client.cpp drive the poll interval; it starts at 0 (poll once immediately), then 5 seconds later, then after that gets doubled after each poll (exponential backoff). This limits network file i/o.

By default, the WorkerListRefreshLimit is 5 minutes, to allow time for scarce workers who are rebooting to appear after a reboot.

I assume that a worker would include its name in the brokerage path for the entire time the worker app is running.  Then clients could use the server status message to know whether a worker is currently available or not. This would minimize network file i/o, and rely on the lighter weight tcp server status messages for the more frequent availability state changes. But based on your other comment, it seems that you are toggling availability more often than the lifetime of the worker app.
Can you explain why this is?
Is it because you limit the number of workers to consider at any given time to 15:
m_WorkerConnectionLimit( 15 )  // default: a maximum of 15 workers simultaneously connected
and want that list to always contain available workers?

Reply:
The backoff strategy of doubling the time between refresh attempts seems reasonable to me. I think maybe we don't need this limit at all (perhaps just a boolean to disable this new behavior, which I think can be on by default.

The current design for workers is that they add themselves to the brokerage when available, and remove themselves when not. They might be unavailable because the FBuildWorker is closed, or they might be unavailable because the machine is under load doing other things (the common case in an office environment where developers run the worked on their machines and the worker works for others when idle.

The aim of removing the token from the brokerage when unavailable is to avoid newly started builds connecting to a worker which will never do any work. The ideal situation is that the workers we connect to are available, so while this doesn't guarantee that, it greatly increases the chances.

Reply:
I think we still need the WorkerListRefreshLimit, for these cases:

1. A job needs a specific worker and the worker never becomes available. This can happen with the worker tags feature. Without a refresh limit, fbuild would hang forever for this case, but I want fbuild to eventually time out.

2. Having a refresh limit would also help the -forceremote case to eventually time out, rather than hang forever.

Thoughts?


I will put the worker available / not available logic back in so that it adds and removes the file on availability.

Reply:
I consider the -forceremote option to be a debug option that should not be used normally. If no workers are available, the build will never finish. There is no mechanism to detect this and time out.

FASTBuild currently assumes that all remote work can be completed locally, so absence of workers would never prevent the build from completing. Do you have a case where the main machine does not have some hardware capability that your workers do that would require some work only ever be done remotely?

If code was added to FASTBuild so the local machine detected when the build was incomplete but some remote-only work was still unfinished, this would require a timeout independent of worker discovery. In-fact, it would have to detect this case even if the Sever was not created (not running in -dist mode). This timeout would in both cases be independent of the period of the build during which we'd attempt to find new workers, so I think this option still isn't needed.

Reply:
Thanks for the explanation.

Yes, I will have future jobs that cannot run locally, examples:
1. A custom compiler that requires a license that is only on one shared machine, not the local machine.
2. Tests that need to be run on a pool of test machines with special hardware; hardware that is not available on the local machine.

I can see how you would prefer the timeout for this to be on the job, rather than the worker refresh. But I could see us needing two different timeouts, the two seem logically and chronologically distinct in the build flow:
a. A worker refresh timeout, with a value for when to stop looking for workers that could meet the needs of these special builds / tests.
b. A job-specific timeout that would apply after successfully connecting to a worker and the job did not complete.

Perhaps you are suggesting that the a. timeout above be moved to some other watchdog thread, not in the worker refresh logic?

Thoughts?